File: 57-insert-interval/insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new ArrayList<>();
        int start = newInterval[0], end = newInterval[1];
        for (int i = 0; i < intervals.length; i++) {
            int newStart = intervals[i][0], newEnd = intervals[i][1];
            if (newEnd < start) {
                res.add(new int[] {newStart, newEnd});
            } else if (end < newStart) {
                res.add(new int[] {start, end});
                res.add(new int[] {newStart, newEnd});
                while (++i < intervals.length) {
                    res.add(intervals[i]);
                }
                return res.toArray(new int[res.size()][]);
            } else {
                start = Math.min(start, newStart);
                end = Math.max(end, newEnd);
            }
        }
        res.add(new int[] {start, end});
        return res.toArray(new int[res.size()][]);
    }
}

File: 57-insert-interval/insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        newStart, newEnd = newInterval[0], newInterval[1]
        intervals.sort(key=lambda x: x[0])
        res = []
        for i in range(len(intervals)):
            currStart, currEnd = intervals[i][0], intervals[i][1]
            # not overlap
            if currEnd < newStart:
                res.append([currStart, currEnd])
            elif currStart > newEnd:
                res.append([newStart, newEnd])
                res += intervals[i:]
                return res
            # overlap
            else:
                newStart = min(currStart, newStart)
                newEnd = max(currEnd, newEnd)
        res.append([newStart, newEnd])
        return res
